crochet-stitcher/src/stages/1-parser.ts

```diff
@@ -1,5 +1,277 @@
+
+enum TokenType {
+    Single = 'sc',
+    Double = 'dc',
+    Slip = 'sl',
+    Chain = 'ch',
+    Triple = 'tr',
+    Increment = 'inc',
+    Decrement = 'dec',
+    Invert = 'inv',
+    Comma = ',',
+    LParen = '(',
+    RParen = ')',
+    Number = 'number',
+    Colon = ':',
+    Asterisk = '*',
+    Colour = 'white',
+    Foundation = 'mr',
+    IntoClause = 'into',
+    Repeat = 'repeat',
+    From = 'from',
+    x = 'x',
+    More = 'more',
+    Times = 'times',
+}
+
+interface ParsedToken {
+    type: TokenType;
+    value: string | number;
+}
+
+function stringTokenLUT(input: string): ParsedToken {
+    // expects strings to be stripped of all punctuation and whitespace
+    input = input.toLowerCase();
+    // Things we understand but do not implement
+    const notImplementedLUT = new Set<string>(['sp', 'hdc', 'slst']);
+    if (notImplementedLUT.has(input)) {
+        throw new TypeError(`${input} is currently not supported!`);
+    }
+
+    // LUT for every valid stitch type
+    // Note that this should also include aliases (eg. ch and chain are both in the LUT)
+    const chainLUT = new Set<string>(['ch', 'chain']);
+    if (chainLUT.has(input)) {
+        return { type: TokenType.Chain, value: input };
+    }
+
+    const singleLUT = new Set<string>(['single', 'sc']);
+    if (singleLUT.has(input)) {
+        return { type: TokenType.Single, value: input };
+    }
+
+    const slipLUT = new Set<string>(['slip', 'sl', 'slipknot']);
+    if (slipLUT.has(input)) {
+        return { type: TokenType.Slip, value: input };
+    }
+
+    const doubleLUT = new Set<string>(['double', 'do']);
+    if (doubleLUT.has(input)) {
+        return { type: TokenType.Double, value: input };
+    }
+
+    const tripleLUT = new Set<string>(['triple', 'tr', 'treble']);
+    if (tripleLUT.has(input)) {
+        return { type: TokenType.Triple, value: input };
+    }
+
+    const incrementLUT = new Set<string>(['increment', 'inc']);
+    if (incrementLUT.has(input)) {
+        return { type: TokenType.Increment, value: input };
+    }
+    const decrementLUT = new Set<string>(['decrement', 'dec']);
+    if (decrementLUT.has(input)) {
+        return { type: TokenType.Decrement, value: input };
+    }
+    const invertLUT = new Set<string>(['invert', 'inv']);
+    if (invertLUT.has(input)) {
+        return { type: TokenType.Invert, value: input };
+    }
+
+    //add more colours here as desired
+    const colourLUT = new Set<string>([
+        'white',
+        'black',
+        'red',
+        'orange',
+        'yellow',
+        'green',
+        'blue',
+        'indigo',
+        'violet',
+    ]);
+    if (colourLUT.has(input)) {
+        return { type: TokenType.Colour, value: input };
+    }
+
+    // Note! if the foundation is specified as a slipknot,
+    // there's no way for us to tell the difference between specifying it
+    // as a stitch vs a foundation knot based on the grammar. As such we'll only catch magic ring
```

Comment:
Some patterns might specify a slip knot as the foundation, but we don't need to support this. This comment is misleading though, since "slip knot" is only a foundation knot, not a valid stitch.